### PR TITLE
[Notifications] Create endpoint for pushing run notifications

### PR DESCRIPTION
**Pull request by roei3000b**
Create `push_notifications` for pushing notifications of specific run.

https://iguazio.atlassian.net/browse/ML-8535

### CODE REVIEW

File: server/py/services/api/api/endpoints/runs.py [hunk 1]
```diff
@@ -518,3 +520,45 @@ async def abort_run(
...
+    framework.utils.notifications.unmask_notification_params_secret_on_task(
+        db, db_session, run
+    )
+    run_notification_pusher_class = (
+        framework.utils.notifications.notification_pusher.RunNotificationPusher
+    )
+    run_notification_pusher_class(
+        [run],
+        run_notification_pusher_class.resolve_notifications_default_params(),
+    ).push()
```
Comment (liranbg) [inline server/py/services/api/api/endpoints/runs.py:582]: isnt it a coroutine you need to await on?

Comment (liranbg) [inline server/py/services/api/api/endpoints/runs.py:582]: this entirely can take some time, move it a background task

**Review by liranbg — CHANGES_REQUESTED** [review]

File: server/py/services/api/api/endpoints/runs.py (hunk 1 above)
Comment (roei3000b) [reply server/py/services/api/api/endpoints/runs.py:582]: No, it isn't
Exactly like here:
https://github.com/mlrun/mlrun/blob/976d1378b9e2ab6bb8b2c41dac5838a47b3321fe/server/py/services/api/main.py#L780-L783

Comment (liranbg) [reply server/py/services/api/api/endpoints/runs.py:582]: then you run blocking IO on your main event loop

File: server/py/services/api/api/endpoints/runs.py [hunk 2]
```diff
@@ -518,3 +520,62 @@ async def abort_run(
...
+        db, db_session, run
+    )
+    run_notification_pusher_class = (
+        framework.utils.notifications.notification_pusher.RunNotificationPusher
+    )
+    run_notification_pusher_object = run_notification_pusher_class(
+        [run],
+        run_notification_pusher_class.resolve_notifications_default_params(),
+    )
+    await run_in_threadpool(run_notification_pusher_object.push)
```
Comment (liranbg) [inline server/py/services/api/api/endpoints/runs.py]: this is all needs to run on a thread
look at the functions implementation, they are not async-awaitable

File: server/py/services/api/main.py [hunk 3]
```diff
@@ -771,16 +771,16 @@ def _push_terminal_run_notifications(
...
+        # self._logger.debug(
+        #     "Got terminal runs with configured notifications", runs_amount=len(runs)
+        # )
+        # run_notification_pusher_class = (
+        #     framework.utils.notifications.notification_pusher.RunNotificationPusher
+        # )
+        # run_notification_pusher_class(
+        #     unmasked_runs,
+        #     run_notification_pusher_class.resolve_notifications_default_params(),
+        # ).push()
```
Comment (liranbg) [inline server/py/services/api/main.py]: on purpose?

**Review by liranbg — CHANGES_REQUESTED** [review]

File: mlrun/utils/notifications/notification_pusher.py [hunk 4]
```diff
@@ -226,15 +226,19 @@ def _should_notify(
...
                         context={
                             "run": run.to_dict(),
                             "notification": notification.to_dict(),
                         },
                     )
-                ) or run_state in ["error", "aborted"]:
+                ) or run_state in [
+                    runtimes_constants.RunStates.error,
+                    runtimes_constants.RunStates.aborted,
+                    runtimes_constants.RunStates.running,
```
Comment (liranbg) [inline mlrun/utils/notifications/notification_pusher.py:240]: why adding running here?

File: server/py/services/api/api/endpoints/runs.py [hunk 5]
```diff
@@ -518,3 +520,62 @@ async def abort_run(
...
+        project,
+        background_tasks,
+        _push_notification,
+        mlrun.mlconf.background_tasks.default_timeouts.push_notifications,
+        framework.utils.background_tasks.BackgroundTaskKinds.push_notification.format(
+            project, uid
+        ),
+        db_session,
+        run,
+    )
```
Comment (liranbg) [inline server/py/services/api/api/endpoints/runs.py]:
```suggestion
    background_task = await run_in_threadpool(
        framework.utils.background_tasks.ProjectBackgroundTasksHandler().create_background_task,
        db_session,
        project,
        background_tasks,
        _push_notification,
        mlrun.mlconf.background_tasks.default_timeouts.push_notifications,
        framework.utils.background_tasks.BackgroundTaskKinds.push_notification.format(
            project, uid
        ),
        db_session,
        run,
    )
    return background_task
```

**Review by liranbg — CHANGES_REQUESTED** [review]
missing ut/integ coverage for using this endpoint

File: mlrun/utils/notifications/notification_pusher.py (hunk 4 above)
Comment (roei3000b) [reply mlrun/utils/notifications/notification_pusher.py:240]: We want to  make running notification also send from the server side.

**Review by liranbg — APPROVED** [review]

**Review by liranbg — APPROVED** [review]

**Pull request merged**